### PR TITLE
Add support for editing the Break / MAB times.

### PR DESCRIPTION
**Pull request by nomis52**

### CODE REVIEW

File: tools/ja-rule/ja-rule.cpp [hunk 1]
```diff
@@ -45,14 +47,22 @@ using ola::io::StdinHandler;
...
 using ola::strings::ToHex;
 using ola::thread::Thread;
+using ola::utils::JoinUInt8;
+using ola::utils::SplitUInt16;
 using std::auto_ptr;
 using std::cout;
 using std::endl;
 using std::string;
 
+DEFINE_string(target_uid, "7a70:00000001", "The UID of the responder.");
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: I think this clashes with the RGB Mixer, which also seems to default to the same values. Either way it (and the RGB Mixer should be listed on http://wiki.openlighting.org/index.php/Open_Lighting_Allocations#UIDs )

File: tools/ja-rule/ja-rule.cpp [hunk 2]
```diff
@@ -45,14 +47,22 @@ using ola::io::StdinHandler;
...
+DEFINE_string(controller_uid, "7a70:00000001", "The UID of the controller.");
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: How can they both default to the same UID?

File: tools/ja-rule/ja-rule.cpp [hunk 3]
```diff
@@ -170,9 +209,43 @@ class MessageHandler : public MessageHandlerInterface {
   void PrintResponse(const Message& message) {
     OLA_INFO << "RC (" << static_cast<int>(message.return_code)
              << "): payload_size: " << message.payload_size;
-    if (message.payload && message.payload_size) {
-      ola::strings::FormatData(&std::cout, message.payload,
-                               message.payload_size);
+
+    if (message.payload_size && message.payload) {
+      rdm_response_code response_code;
+      // Ignore the start code.
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: Add a todo to ensure it's the RDM start code, or is that done uC side?

File: tools/ja-rule/ja-rule.cpp [hunk 4]
```diff
@@ -170,9 +209,43 @@ class MessageHandler : public MessageHandlerInterface {
...
+               << "): payload_size: " << message.payload_size;
+      return;
+    }
+
+    if (message.payload_size == sizeof(time)) {
+      time = JoinUInt8(message.payload[1], message.payload[0]);
+      if (format == MICRO_SECONDS) {
+        OLA_INFO << "Time: " << time << " uS";
+      } else if (format == TENTHS_OF_MILLI_SECONDS) {
+        float adjusted_time = time / 10.0;
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: You need to divide by 100 to get micro's from 0.1 millis.

File: tools/ja-rule/ja-rule.cpp [hunk 5]
```diff
@@ -214,12 +294,35 @@ class InputHandler {
 
   void Input(int c) {
     switch (c) {
+      case 10:  // Enter
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: \n is probably more obvious isn't it?

File: tools/ja-rule/ja-rule.cpp [hunk 6]
```diff
@@ -250,6 +368,30 @@ class InputHandler {
       case 'w':
         WriteLog();
         break;
+      case 'x':
+        GetRDMWaitTime();
+        break;
+      case 'X':
+        cout << "Editing RDM Wait time, use +/- to adjust, Enter commits, "
+             << " Esc to abort" << endl;
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: Double space across the two lines.

File: tools/ja-rule/ja-rule.cpp [hunk 7]
```diff
@@ -250,6 +368,30 @@ class InputHandler {
...
+        cout << "Editing MAB, use +/- to adjust, Enter commits, Esc to abort"
+             << endl;
+        m_mode = EDIT_MAB;
+        break;
+      case 'z':
+        GetRDMBroadcastListen();
+        break;
+      case 'Z':
+        cout << "Editing RDM Broadcast Listen Time, use +/- to adjust, Enter "
+             << " commits, Esc to abort" << endl;
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: Again.

File: tools/ja-rule/ja-rule.cpp [hunk 8]
```diff
@@ -259,26 +401,54 @@ class InputHandler {
     cout << "Commands:" << endl;
     cout << " 0 - Send a 0 length DMX frame" << endl;
     cout << " 2 - Send 2 DMX frames back to back" << endl;
+    cout << " b - Get Break time" << endl;
+    cout << " B - Set Break time" << endl;
     cout << " d - Send a DUB frame" << endl;
     cout << " e - Send Echo command" << endl;
     cout << " f - Fetch Flags State" << endl;
     cout << " h - Print this help message" << endl;
+    cout << " i - Identify On to --target-uid" << endl;
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: You could display the target UID too.

File: tools/ja-rule/ja-rule.cpp [hunk 9]
```diff
@@ -259,26 +401,54 @@ class InputHandler {
...
+    cout << " I - Identify Off to --target-uid" << endl;
     cout << " l - Fetch Logs" << endl;
+    cout << " m - Send a broadcast mute" << endl;
+    cout << " M - Send a mute to the device specified in --target-uid" << endl;
     cout << " q - Quit" << endl;
     cout << " r - Reset" << endl;
     cout << " t - Send DMX frame" << endl;
-    cout << " u - Send an UnMute-all frame" << endl;
+    cout << " u - Send an broadcast unmute" << endl;
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: SPaG a

File: tools/ja-rule/ja-rule.cpp [hunk 10]
```diff
@@ -319,6 +489,70 @@ class InputHandler {
...
+        m_mab--;
+      }
+      cout << "MAB is now " << m_mab << " uS" << endl;
+    } else if (m_mode == EDIT_RDM_WAIT_TIME) {
+      if (increase) {
+        m_rdm_wait_time++;
+      } else {
+        m_rdm_wait_time--;
+      }
+      float wait_time = m_rdm_wait_time / 10.0;
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: Again 100.

File: tools/ja-rule/ja-rule.cpp [hunk 11]
```diff
@@ -170,9 +209,43 @@ class MessageHandler : public MessageHandlerInterface {
...
+                                   message.payload_size);
+        }
+      } else {
+        OLA_INFO << *response;
+      }
+    }
+  }
+
+  void PrintTime(const Message& message, TimeFormat format) {
+    uint16_t time = 0;
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: Is it worth using an unreserved word? It's certainly confusing GitHub's formatting.

File: tools/ja-rule/ja-rule.cpp (hunk 10 above)
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: Actually I don't think this is, is it.

File: tools/ja-rule/ja-rule.cpp [hunk 12]
```diff
@@ -319,6 +489,70 @@ class InputHandler {
...
+      }
+      float listen_time = m_rdm_broadcast_listen / 10.0;
+      cout << "RDM Broadcast listen is now " << listen_time << " mS"
+           << endl;
+    }
+  }
+
+  void Commit() {
+    if (!CheckForDevice()) {
+      return;
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: Throw an error message?

File: tools/ja-rule/ja-rule.cpp [hunk 13]
```diff
@@ -319,6 +489,70 @@ class InputHandler {
...
+    }
+
+    if (m_mode == EDIT_BREAK) {
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: Switch case?

File: tools/ja-rule/ja-rule.cpp (hunk 12 above)
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: Actually should be done in CheckForDevice if it isn't already.

**Comment by peternewman** [comment]
Is it worth having a SendMessage that's doesn't require data and length, and doesn't complain that data is NULL etc, for param less gets?

File: tools/ja-rule/ja-rule.cpp (hunk 2 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: I've reserved a range for the controller UIDs in the webpage.

File: tools/ja-rule/ja-rule.cpp (hunk 1 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: The responder UID can be anything, when I use this I set --target_uid to the device I'm trying to control.

I thought about making the --target-uid param required, but you should be able to use the tool without specifying the RDM parameters.

File: tools/ja-rule/ja-rule.cpp (hunk 3 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Done

File: tools/ja-rule/ja-rule.cpp (hunk 4 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Changed to ms.

File: tools/ja-rule/ja-rule.cpp (hunk 5 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Fixed.

File: tools/ja-rule/ja-rule.cpp (hunk 6 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Fixed.

File: tools/ja-rule/ja-rule.cpp (hunk 7 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Fixed

File: tools/ja-rule/ja-rule.cpp (hunk 8 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Done

File: tools/ja-rule/ja-rule.cpp (hunk 9 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Fixed.

File: tools/ja-rule/ja-rule.cpp (hunk 10 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Yup, this is in ms so  / 10.0 is correct.

File: tools/ja-rule/ja-rule.cpp (hunk 11 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Done

File: tools/ja-rule/ja-rule.cpp (hunk 12 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: It's done in CheckForDevice

File: tools/ja-rule/ja-rule.cpp (hunk 13 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: You can't define variables in a switch.

**Comment by nomis52** [comment]
> Is it worth having a SendMessage that's doesn't require data and length, and doesn't complain that data is NULL etc, for param less gets?

Maybe later I can override the method. Right now I'm just trying to get it working.

File: tools/ja-rule/ja-rule.cpp (hunk 1 above)
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: Okay. Perhaps it's worth clarifying the flag; "The UID of the responder to control"?

File: tools/ja-rule/ja-rule.cpp [hunk 14]
```diff
@@ -407,8 +709,20 @@ class InputHandler {
 int main(int argc, char **argv) {
   ola::AppInit(&argc, argv, "[ options ]", "Ja Rule Admin Tool");
 
+  auto_ptr<UID> target_uid(UID::FromString(FLAGS_target_uid));
+  if (!target_uid.get()) {
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: This should probably check or at least warn that the UID isn't a broadcast one, or do you want to be able to test odd stuff?

File: tools/ja-rule/ja-rule.cpp [hunk 15]
```diff
@@ -407,8 +709,20 @@ class InputHandler {
...
+    OLA_WARN << "Invalid Target UID: '" << FLAGS_target_uid << "'";
+    exit(ola::EXIT_USAGE);
+  }
+
+  auto_ptr<UID> controller_uid(UID::FromString(FLAGS_controller_uid));
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: Again.

File: tools/ja-rule/ja-rule.cpp [hunk 16]
```diff
@@ -170,9 +210,46 @@ class MessageHandler : public MessageHandlerInterface {
...
+    if (message.return_code != 0) {
+      OLA_INFO << "Failed (" << static_cast<int>(message.return_code)
+               << "): payload_size: " << message.payload_size;
+      return;
+    }
+
+    if (message.payload_size == sizeof(value)) {
+      value = JoinUInt8(message.payload[1], message.payload[0]);
+      if (format == MICRO_SECONDS) {
+        OLA_INFO << "Time: " << value << " uS";
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: SPaG us for micro-seconds

File: tools/ja-rule/ja-rule.cpp [hunk 17]
```diff
@@ -170,9 +210,46 @@ class MessageHandler : public MessageHandlerInterface {
...
+      } else if (format == TENTHS_OF_MILLI_SECONDS) {
+        float adjusted_time = value / 10.0;
+        OLA_INFO << "Time: " << adjusted_time << " mS";
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: ms

File: tools/ja-rule/ja-rule.cpp [hunk 18]
```diff
@@ -319,6 +492,74 @@ class InputHandler {
...
   }
 
+  void Adjust(bool increase) {
+    if (m_mode == EDIT_BREAK) {
+      if (increase) {
+        m_break++;
+      } else {
+        m_break--;
+      }
+      cout << "Break is now " << m_break << " uS" << endl;
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: us

File: tools/ja-rule/ja-rule.cpp [hunk 19]
```diff
@@ -319,6 +492,74 @@ class InputHandler {
...
+    } else if (m_mode == EDIT_MAB) {
+      if (increase) {
+        m_mab++;
+      } else {
+        m_mab--;
+      }
+      cout << "MAB is now " << m_mab << " uS" << endl;
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: us

File: tools/ja-rule/ja-rule.cpp [hunk 20]
```diff
@@ -319,6 +492,74 @@ class InputHandler {
...
+    } else if (m_mode == EDIT_RDM_WAIT_TIME) {
+      if (increase) {
+        m_rdm_wait_time++;
+      } else {
+        m_rdm_wait_time--;
+      }
+      float wait_time = m_rdm_wait_time / 10.0;
+      cout << "RDM Wait time is now " << wait_time << " mS" << endl;
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: SPaG ms

File: tools/ja-rule/ja-rule.cpp [hunk 21]
```diff
@@ -319,6 +492,74 @@ class InputHandler {
...
+    } else if (m_mode == EDIT_RDM_BROADCAST_LISTEN) {
+      if (increase) {
+        m_rdm_broadcast_listen++;
+      } else {
+        m_rdm_broadcast_listen--;
+      }
+      float listen_time = m_rdm_broadcast_listen / 10.0;
+      cout << "RDM Broadcast listen is now " << listen_time << " mS"
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: ms

**Comment by peternewman** [comment]
LGTM, just a few more comments.

File: tools/ja-rule/ja-rule.cpp (hunk 1 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Done

File: tools/ja-rule/ja-rule.cpp (hunk 16 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Done

File: tools/ja-rule/ja-rule.cpp (hunk 17 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Done

File: tools/ja-rule/ja-rule.cpp (hunk 18 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Done

File: tools/ja-rule/ja-rule.cpp (hunk 19 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Done

File: tools/ja-rule/ja-rule.cpp (hunk 20 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Done

File: tools/ja-rule/ja-rule.cpp (hunk 21 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Done

File: tools/ja-rule/ja-rule.cpp (hunk 14 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Broadcast is useful. This is a dev tools so there are no gloves :)

File: tools/ja-rule/ja-rule.cpp (hunk 15 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Done

File: tools/ja-rule/ja-rule.cpp (hunk 14 above)
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: Yep. Hence I wonder if it should just be a warning to STDOUT that doesn't affect the processing. "Are you sure you meant to do that?"

**Comment by peternewman** [comment]
LGTM

File: tools/ja-rule/ja-rule.cpp (hunk 14 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Again, it's assumed if you're using this you know what you're doing. I'm trying to limit the tool to printing useful information.

Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: Okay, understood.

**Pull request merged**